src/components/ComboButton/ComboButton.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * @file Combo button.
+ * @copyright IBM Security 2019
+ */
+
+import classnames from 'classnames';
+import PropTypes from 'prop-types';
+import React, { useRef, useState } from 'react';
+import { ChevronDown16, ChevronUp16 } from '@carbon/icons-react';
+import { settings } from 'carbon-components';
+
+import Button from '../Button';
+import OverflowMenu from '../OverflowMenu';
+import { TooltipDirection } from '../IconButton/IconButton';
+
+import { namespace as buttonNamespace } from '../Button/Button';
+
+import { getComponentNamespace } from '../../globals/namespace';
+
+export const namespace = getComponentNamespace('combo-button');
+
+const { prefix } = settings;
+
+const ComboButton = ({ children, className, direction, primaryAction }) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const wrapper = useRef(null);
+
+  const getMenuOffset = () => {
+    const { top } = wrapper.current.getBoundingClientRect();
+    const isTop = direction === TooltipDirection.TOP;
+    return {
+      top: isTop ? top : top * -1,
+      left: 'auto',
+    };
+  };
+
+  return (
+    <div
+      className={classnames(namespace, className)}
+      ref={wrapper}
+      data-floating-menu-container
+    >
+      <Button
+        {...primaryAction}
+        className={`${namespace}__button`}
+        renderIcon={
+          children && children.length > 1 ? null : primaryAction.renderIcon
+        }
+      >
+        {primaryAction.label}
+      </Button>
+
+      {children && (
+        <OverflowMenu
+          className={classnames(
+            // Button-specific classes:
+            buttonNamespace,
+            `${prefix}--btn`,
+            `${prefix}--btn--primary`,
+
+            // Button as a child of combo button:
+            `${namespace}__button`,
+
+            // Overflow menu as a child of combo button:
+            `${namespace}__overflow-menu`
+          )}
+          direction={direction}
+          menuOffset={getMenuOffset}
+          menuOptionsClass={`${prefix}--list-box__menu`}
+          onClose={() => setIsOpen(false)}
+          onOpen={() => setIsOpen(true)}
+          renderIcon={isOpen ? ChevronUp16 : ChevronDown16}
+        >
+          {children}
+        </OverflowMenu>
+      )}
+    </div>
+  );
+};
+
+ComboButton.propTypes = {
+  /** @type {Object.<string, string>} An object list of props for the primary button. */
+  primaryAction: PropTypes.shape({
```

Comment:
I was thinking that the primary action would be gathered from the children group. For example usage like this:

```jsx
<ComboButton>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 1
  </ComboButtonItem>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 2
  </ComboButtonItem>
</ComboButton>
```
Would take the first action and make it the primary action, basically abstracting away that logic within the component. Alternatively, a consumer can tag one of the `ComboButtonItem`s as the primary action:

```jsx
<ComboButtonItem renderIcon={Icon} onClick={() => {}} primary>Action 1</ComboButtonItem>
```

What do you think? Is this something that is doable and makes sense? My main motivation is to focus on flat props (instead of objects) which are easier to work with and debug IMO.

Reply:
Ah yeah that makes sense! 👍 Sounds good to me; I can make that change

Reply:
Well @diego-codes I was thinking about this & it may be difficult to achieve 🤔 

Right now `ComboButtonItem` is just a renamed `OverflowMenuItem` (side note: doesn't *have* to be this way, but it was definitely the path of least resistance, otherwise I was getting strange focus / tabbing results that I messaged you in Slack about).

The props for `ComboButtonItem` are pretty different from the props for `Button` (which is the underlying "primary action"). I notice in your mocked example you have button props like `renderIcon` etc but those aren't available to `ComboButtonItem` currently.

To that end, I think the biggest challenge is the `ComboButtonItem` / `OverflowMenuItem` components' `itemText` prop. That prop accepts nodes / anything really, so the user has a lot of freedom there. In the story examples I tried explaining how best to use `itemText` to get a well-formatted `ComboButton` in general, but it would be strange to take what's passed into `itemText`, especially when we are talking about maybe `span` elements and icon components, and then fit that into a `Button`. 🤔 

Anyway what do you think about all of this? 
I'm thinking that the 2 routes to solve this, if we want to stick with this general structure, are:
1. Keep `primaryActions` object
2. Have something like `ComboButtonAction` that renames a `Button` for this purpose, so it can be used as a sibling of `ComboButtonItem` components? Or just have a `Button` be used directly? My one worry here is that the API may be too open / lenient, it may imply that it's okay to put any number of other children here. One way to get around may be to add some warning if something other than `ComboButtonItem` or `ComboButtonAction`/`Button` are added as children of `ComboButton`?

What do you think? 🤔 😅

Reply:
I definitely see what you mean. I like the idea of having a separate component for the primary action. It reminds me of the Carbon UI Shell and DataTable components where it allows developers compose components with the provided building blocks.

I'm also thinking that another approach could be that the `ComboButtonItem` component becomes a shell component that renders either a `Button` or `OverflowMenu`. Pardon my over directing, here is what I could see working:

```jsx
const ComboButtonItem = props => {
  const { primary, renderIcon: Icon, children, className, onClick, iconDescription } = props;
  const classNames = classnames(namespace, { [`${namespace}--primary`]: primary }, className);
  if (primary) {
    return (
      <Button
        className={classNames}
        renderIcon={Icon}
        onClick={onClick}
        iconDescription={iconDescription}
      >
        {children}
      </Button>
    );
  }

  return (
    <OverflowMenuItem
      className={classNames}
      itemText={
        <Fragment>
          <span title={children}>{children}</span>
          <Icon iconDescription={iconDescription} />
        </Fragment>
      }
    />
  );
};

ComboButtonItem.propTypes = {
  primary: PropTypes.bool,
  className: PropTypes.string,
  renderIcon: PropTypes.node,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  iconDescription: PropTypes.string,
};
```

This is just a hunch that it would work, but I haven't tested it. Either approach I think would work.

Reply:
Thanks @diego-codes! This definitely makes sense & I appreciate you sharing your approach. 

One challenge that I see is that the primary `Button` and the `OverflowMenuItem` components are not siblings within `children` of the `ComboButton` component. 🤔 The `Button` needs to be outside of the underlying `OverflowMenu`,  whereas the `OverflowMenuItem` components need to be inside... I'm just wondering how best to accomplish that without adding too much complexity.